### PR TITLE
OneTBB NUMA support API spec

### PR DESCRIPTION
**Pull request by ivankochin**
This patch contains the oneTBB NUMA support API specification.

@akukanov, @alexey-katranov, @aleksei-fedotov please review if possible.

### CODE REVIEW

**Comment by rscohn2** [comment]
Ignore the build error. I am in the middle of redoing the publish process.

Is this intended to go into the 1.0 spec?

By the way, I have been looking at how nvidia supports multiple cards and the ability to create threads in specific numa nodes will be useful. You may want to create threads in a subset of the numa nodes (the ones with separate PCI interfaces and GPU cards attached). Your API will support it in a natural way.

**Comment by alexey-katranov** [comment]
> Is this intended to go into the 1.0 spec?

Yes, we just noticed that we missed this API in the Spec. 

As for additional use cases, we are thinking about other criteria for constraints. I believe we can extend the API with new use cases in future.

File: source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst [hunk 1]
```diff
@@ -109,6 +134,22 @@ Member functions
...
 
+.. cpp:function:: task_arena(constraints a_constraints, unsigned reserved_for_masters = 1, priority a_priority = priority::normal)
+
+    Creates a ``task_arena`` with a certain constraints(``a_constraints``) and priority
+    (``a_priority``).  Some portion of the limit can be reserved for application threads with
+    ``reserved_for_masters``.  The amount for reservation cannot exceed the limit.
+
+    .. caution::
+
+        If ``max_concurrency`` and ``reserved_for_masters`` are
```
Comment (aleksei-fedotov) [inline source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst:145]:
```suggestion
        If ``constraints::max_concurrency`` and ``reserved_for_masters`` are
```
?

File: source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst [hunk 2]
```diff
@@ -109,6 +134,22 @@ Member functions
...
+        explicitly set to be equal and greater than 1, oneTBB worker threads will never
+        join the arena. As a result, the execution guarantee for enqueued tasks is not valid
+        in such arena. Do not use ``task_arena::enqueue()`` with an arena set to have no worker threads.
+
+    If ``constraints.numa_node`` is specified, then all threads that enter the arena are automatically
```
Comment (aleksei-fedotov) [inline source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst:150]:
```suggestion
    If ``constraints::numa_node`` is specified, then all threads that enter the arena are automatically
```
Qualify via double colon?

**Comment by rscohn2** [comment]
Yes. API that generate and accept vectors of numa nodes should be a good fit for other use cases.

File: source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst [hunk 3]
```diff
@@ -202,6 +247,48 @@ Member functions
         Any number of threads outside of the arena can submit work to the arena and be blocked.
         However, only the maximal number of threads specified for the arena can participate in executing the work.
 
+Example
+-------
+
+The example demonstrates task_arena NUMA support extensions. Each constructed task_arena is pinned
+to the corresponding node. If the platform does not have NUMA capabilities, the example will create
+only one ``task_arena`` and threads entering this ``task_arena`` will not be bound to any hardware
+resources.
```
Comment (aleksei-fedotov) [inline source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst:256]: I think we don't need to clarify too precisely what is happening here. This is the spec. We only need to tell that this is an example of how work can be executed on different NUMA nodes simultaneously.

**Review by aleksei-fedotov — CHANGES_REQUESTED** [review]

File: source/elements/oneTBB/source/info_namespace.rst [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            int default_concurrency(numa_node_id id = tbb::task_arena::automatic);
+        }
+    }
+
+A tbb::info namespace contains interfaces for getting library configuration information.
+
+Types
+-----
+
+``numa_node_id`` - Represents the certain NUMA node identifier.
```
[the remark this answers is not in the document]
Comment (ivankochin) [reply source/elements/oneTBB/source/info_namespace.rst:25]: I agree with you about "the" article, but I am prefer the following form:

```suggestion
``numa_node_id`` - Represents NUMA node identifier.
```

File: source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst [hunk 5]
```diff
@@ -109,6 +134,22 @@ Member functions
         join the arena. As a result, the execution guarantee for enqueued tasks is not valid
         in such arena. Do not use ``task_arena::enqueue()`` with an arena set to have no worker threads.
 
+.. cpp:function:: task_arena(constraints a_constraints, unsigned reserved_for_masters = 1, priority a_priority = priority::normal)
+
+    Creates a ``task_arena`` with a certain constraints(``a_constraints``) and priority
+    (``a_priority``).  Some portion of the limit can be reserved for application threads with
+    ``reserved_for_masters``.  The amount for reservation cannot exceed the limit.
```
[the remark this answers is not in the document]
Comment (ivankochin) [reply source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst:141]: This formulation was copied from the previous constructor, so I would prefer it to save consistency.

File: source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst [hunk 6]
```diff
@@ -202,6 +247,46 @@ Member functions
         Any number of threads outside of the arena can submit work to the arena and be blocked.
         However, only the maximal number of threads specified for the arena can participate in executing the work.
 
+Example
+-------
+
+The example demonstrates task_arena NUMA support extensions. Each constructed task_arena is pinned
+to the corresponding NUMA node.
```
Comment (ivankochin) [inline source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst:254]:
```suggestion
The example demonstrates ``task_arena`` NUMA support API. Each constructed task_arena is pinned
to the corresponding NUMA node.
```

File: source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst [hunk 7]
```diff
@@ -202,8 +247,47 @@ Member functions
         Any number of threads outside of the arena can submit work to the arena and be blocked.
         However, only the maximal number of threads specified for the arena can participate in executing the work.
 
+Example
+-------
+
+The example demonstrates ``task_arena`` NUMA support API. Each constructed task_arena is pinned
```
Comment (ivankochin) [inline source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst:253]:
```suggestion
The example demonstrates ``task_arena`` NUMA support API. Each constructed ``task_arena`` is pinned
```

Comment (aleksei-fedotov) [inline source/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.rst:253]:
```suggestion
The example demonstrates ``task_arena`` NUMA support API. Each constructed ``task_arena`` is pinned
```

**Review by aleksei-fedotov — APPROVED** [review]

**Comment by ivankochin** [comment]
@rscohn2 , we already reviewed this PR, please, merge this patch is you have no objections.

**Pull request merged**